src/Perpetuum/Zones/NpcSystem/NPCBossInfo.cs

```diff
@@ -210,6 +210,14 @@ public override bool Equals(object obj)
 
         public bool Equals(NpcBossInfo other)
         {
+            if (ReferenceEquals(this, other))
+            {
+                return true;
+            }
+            else if (other == null)
+            {
+                return false;
+            }
             return other._id == _id && other.FlockId == FlockId;
```

Comment:
Always do null checks first. That way you can guarantee your object is not null for the rest of your method :)
`return other != null && (ReferenceEquals(this, other) || (other._id == _id && other.FlockId == FlockId));`

Reply:
Also, *strictly speaking* the `ReferenceEquals` will do a null check implicitly because `this` can't be null. *However*, we wouldn't know if it's because null or because different, so we still have to do an explicit check.

Reply:
Also, just for completeness, `ReferenceEquals` and `==` do the same job *normally*. However, you can overload the `==operator` which is why I prefer to use `ReferenceEquals` explicitly when I'm sure I only want a reference check.